fenicsadapter/fenicsadapter.py

```diff
@@ -55,64 +65,65 @@ def __init__(self, adapter_config_filename='precice-adapter-config.json'):
 
         # Solver state used by the Adapter internally to handle checkpointing
         self._checkpoint = None
 
-        self._dss = None  # measure for boundary integral
-
         # Dirichlet boundary for FSI Simulations
         self._Dirichlet_Boundary = None  # stores a dirichlet boundary (if provided)
 
-        # Necessary flag in checkpoint storing function
+        # Necessary bools for enforcing proper control flow / warnings to user
         self._first_advance_done = False
-
-        # Flag to see if 2D - 3D coupling needs to be applied
         self._apply_2d_3d_coupling = False
+        self._non_standard_initialization = False
 
     def create_coupling_expression(self, data=None):
         """
-        Creates an object of class GeneralInterpolationExpression or ExactInterpolationExpression which does
-        not carry any data. The adapter will hold this object till the coupling is on going.
+        Creates a FEniCS Expression in the form of an object of class GeneralInterpolationExpression or
+        ExactInterpolationExpression. The adapter will hold this object till the coupling is on going.
 
         Parameters
         ----------
-        data : numpy.ndarray
-            The coordinates of the vertices. Coordinates of vertices are stored in a
-            numpy array [N x D] where N = number of vertices and D = dimensions of geometry
+        data : array_like
+            The coupling data. A numpy array [N x D] where N = number of vertices and D = dimensions of geometry.
+            If no data is provided then a numpy array of shape (N, D) with zero values is used.
 
         Returns
         -------
-        coupling_expression : object of FEniCS class CustomExpression
-            Reference to object of FEniCS class CustomExpression.
+        coupling_expression : Object of class dolfin.functions.expression.Expression
+            Reference to object of class GeneralInterpolationExpression or ExactInterpolationExpression.
         """
+        if self._non_standard_initialization and data is None:
+            warn("initialize_data() was called but the data was not supplied for creating the coupling expression")
+
         try:  # works with dolfin 1.6.0
             # element information must be provided, else DOLFIN assumes scalar function
             coupling_expression = self._my_expression(element=self._function_space.ufl_element())
         except (TypeError, KeyError):  # works with dolfin 2017.2.0
             coupling_expression = self._my_expression(element=self._function_space.ufl_element(), degree=0)
 
-        if not data:
-            data = np.zeros((self._n_vertices, self._fenics_dimensions))
+        n_vertices, _ = self._coupling_mesh_vertices.shape
+
+        if data is None:
+            # standard initialization with zero valued data
+            data = np.zeros((n_vertices, self._fenics_dimensions))
```

Comment:
This would **always** create data for vector valued coupling functions. I think we should additionally check the `self._read_function_type`, if we want to treat this case rigurously.